### ISSUE TITLE
Separate vscode-java version from development target

### ISSUE THREAD

**Issue by ericwill** [eclipse-che/che#16851]
# The Problem
* JDT-LS will only run on Java 11 starting with Eclipse 2020-09, so probably in July 2020 already
* Currently in Che, developing Java code using the Java plug-in means your code runs on the same version as JDT-LS does
* Bumping the min version for JDT-LS to Java 11 means no Java 8 development can happen in Che
* It would also mean that as Java versions continue to be released, we'd need a new plug-ins, and also as the JDT-LS minimum version increases we'd need more plug-ins
    * This becomes complicated and messy to maintain

# Solutions
We need a way to separate the Java tooling environment from the Java target development environment/compliance level. Investigation needs to be done first, but an ideal solution would allow:

* Users to bring their own JDK, either in the form of a container with an entry point, or something similar
* Legacy Java version flow to be supported, so that users can develop against older versions of Java while still using the latest tooling
* Allow users to bring their own custom JDKs as a target environment/compliance level
* An approach where we only have to maintain one central Java plugin

Subtasks/issues to follow after suitable investigation into potential designs has been conducted.

Sub-issues:
* [x] Support object array preferences in a devfile #17481
* [ ] Volume sharing between init container and che-plugin container does not working #17102
* [ ] Provide experimental "portable JDK" devfile #17924

**Comment by tsmaeder** [comment]
# Separation of Java runtime and target

Currently, we have no possiblity in Che to run the Java language server included with VSCode-Java (jdt.ls) with a different JDK than the one jdt.ls runs on itself. This is a problem for a number of reasons:

1. Pretty soon, jdt.ls will require Java 11 to run. This would meant that we cannot provide a devfile that will use Java 8 to provide language services for Java 8 projects. For example, when looking up the source of java.io.File in the debugger, it will always present the user with the source from Java 11, not Java 8. 

2. Currently, we need to maintain a separate plugin for each of the Java versions we want to support. The problem is worsened by the fact that we have multiple plugins that include Java support, in particular Quarkus. All these versions need to be maintainted.

3. Even if we had versions of our plugins for each major Java version, we would be overwhelmed with maintaining a plugin for each Java version published. One of the goals of developing with containers is to prevent the "it works on my machine"-problem. So ideally, jdt.ls would see the exact same version of the JDK the user will use to run their program in production.

# Goals

1. Run jdt.ls with a different JDK than the target JDK
   This is non-optional becuase jdt.ls needs to work against Java 8, while running on Java >= 11

2. Provide only one version of the Java plugin and still support multiple JDK target versions

3. Use a user-provided target JDK

# Approaches

jdt.ls has the possiblity to use a different JDK as the target than the one it's running on. A list of jdk install paths can be passed as a VS Code-Java preference. We can provide preference defaults for Theia in a devfile.

So the problem becomes one of presenting a second JDK to VS Code Java in a location on the file system. 

## 1. Include the JRE with the plugin

The trivial solution to the problem is to include a target and a runtime Java version in the Java plugins that we already have. We'd use the same runtime (Java 11) for jdt.ls and add a second jdk to serve as a target JDK. This addresses goal 1, but not goals 2 and 3.

## 2. Run jdt.ls in the "dev" container

Theoretically, we could also run jdt.ls on the user-provided "dev container": however, that would still require two JDK containers to be present on the dev-container, one to run jdt.ls and one as a target.

## 3. Mount the JDK into the container

Beyond that, we would have to provide the JDK to jdt.ls from the outside. We can separate this issue into two parts: first, the JDK files need to be provided somehow, and second, the file need to be consumed by VS Code Java, and ultimately jdt.ls.

For the consumption side, @benoitf has suggested we might try to overrride the default file system provider for the Java. There might then be a local server container in the pod that could serve the necessary files via some protocol. 

The only other way seems to be via a kubernetes volume mounted in the Java plugin container. 

One idea was to run a local file server in a container the pod that would server a JDK directory installed in that container and to mount that JDK inside the Java plugin sidecar. However, I don't believe that Kubernetes and/or OpenShift would allow that in a container.    

The next approach would be to provide the JDK in a kubernetes volume. Option one would be a non-persistent volume. This would mean we have to put the JDK into that volume each time. I did a quick test on openshift.io copying the jdk 11 directory from /usr/lib/jvm into /projects (ca. 250mb) : when using ephemeral storage, the copy took <1.5 seconds. When using persistent storage, the copy took ~15s.

That measurement raises the question whether it's worth using any persistent storage for the JDK considering the performance of the persistent storage is likely to be worse than what can be expected from ephemeral storage.

# Proposal

So my proposal would be to package JDK versions as init containers. When the init container is runs, it will copy the included JDK to a directory given as an env variable. A devfile for a particular version of Java would then point jdt.ls to a "target runtime" by including the proper VS Code-Java settings. I would leave the choice of using either persistent or ephemeral storage up to the devfile. I believe currently, there is no way to specify a volume as "ephemeral" or "local": that can only be done globally for the devfile.

**Comment by benoitf** [comment]
I think the idea is to be able to use the existing java images available on docker registries [1] (and not to package and maintain ourselves images)

For example https://hub.docker.com/_/openjdk?tab=tags

so if proposal is working fine with these images

virtual filesystem approach would have been cool to me because you could always shared that to many workspaces. Less environment variables, more automatic discovery.

**Comment by tsmaeder** [comment]
@benoitf we are already capable of overriding container entrypoints: it's just a matter of doing a "cp $SRC $DEST in the container. This mechanism would and should not be restricted to Java. 
I'm pretty sure it would work on a kubernetes level, what remains to be seen is how we can express it in a devfile (some work might be required).

**Comment by tsmaeder** [comment]
So here's a devfile that sets up openj9 as the target jdk:

```
metadata:
  name: testJDKVersionKube
projects:
  - name: console-java-simple
    source:
      location: 'https://github.com/che-samples/console-java-simple.git'
      type: git
      branch: java1.11
attributes:
  persistVolumes: 'false'
components:
  - id: redhat/java11/latest
    type: chePlugin
  - mountSources: true
    memoryLimit: 512Mi
    type: dockerimage
    volumes:
      - name: jdk
        containerPath: /tmp/jdk
      - name: m2
        containerPath: /home/user/.m2
    image: 'quay.io/eclipse/che-java11-maven:7.13.1'
    alias: maven
    env:
      - value: ''
        name: MAVEN_CONFIG
      - value: >-
          -XX:MaxRAMPercentage=50 -XX:+UseParallelGC -XX:MinHeapFreeRatio=10
          -XX:MaxHeapFreeRatio=20 -XX:GCTimeRatio=4
          -XX:AdaptiveSizePolicyWeight=90 -Dsun.zip.disableMemoryMapping=true
          -Xms20m -Djava.security.egd=file:/dev/./urandom
        name: JAVA_OPTS
      - value: >-
          -XX:MaxRAMPercentage=50 -XX:+UseParallelGC -XX:MinHeapFreeRatio=10
          -XX:MaxHeapFreeRatio=20 -XX:GCTimeRatio=4
          -XX:AdaptiveSizePolicyWeight=90 -Dsun.zip.disableMemoryMapping=true
          -Xms20m -Djava.security.egd=file:/dev/./urandom -Duser.home=/home/user
        name: MAVEN_OPTS
      - value: >-
          -XX:MaxRAMPercentage=50 -XX:+UseParallelGC -XX:MinHeapFreeRatio=10
          -XX:MaxHeapFreeRatio=20 -XX:GCTimeRatio=4
          -XX:AdaptiveSizePolicyWeight=90 -Dsun.zip.disableMemoryMapping=true
          -Xms20m -Djava.security.egd=file:/dev/./urandom
        name: JAVA_TOOL_OPTIONS
  - referenceContent: |
      apiVersion: v1
      kind: Pod
      metadata:
        name: ws
      spec:
        initContainers:
          - name: init-myservice
            image: adoptopenjdk/openjdk11-openj9
            command: ['sh', '-c', "echo starting && cp -R /opt/java/openjdk/ /tmp/jdk && echo done && sleep 20"]
            volumeMounts: 
              - name:  remote-endpoint
                mountPath: /tmp/jdk
    type: kubernetes
apiVersion: 1.0.0
commands:
  - name: maven build
    actions:
      - workdir: '${CHE_PROJECTS_ROOT}/console-java-simple'
        type: exec
        command: mvn clean install
        component: maven
  - name: maven build and run
    actions:
      - workdir: '${CHE_PROJECTS_ROOT}/console-java-simple'
        type: exec
        command: mvn clean install && java -jar ./target/*.jar
        component: maven
```

In addition, you'll have to add the following to the user theia settings:

```
"java.configuration.runtimes": [
    {
        "name": "JavaSE-11",
        "path": "/remote-endpoint/openjdk",
        "default": true
    }
```

**Comment by tsmaeder** [comment]
A little demo: https://youtu.be/NLuG7gWdL6w

**Comment by tsmaeder** [comment]
There are two problems currently: 
* We have no way to share ephemeral volumes between init containers and plugins: https://github.com/eclipse/che/issues/17102
* We can't set the jdk preference for jdt.ls properly: https://github.com/eclipse/che/issues/16594  indicates that some work has been done for allowing arrays of simple values, but we need arrays of objects: basically, a general solution.

**Comment by tsmaeder** [comment]
Note that in the PoC, I use a stock jdk image. Also, we could use the maven container as  an init container, as well: pulling it twice should not be expensive.

**Comment by l0rd** [comment]
The init container solution solves nicely the goal n.3: "Use a user-provided target JDK".

But I am wondering if, for the target JDKs that we are going to include in Che (Because we are still going to provide some default target JDK right?) this is a better solution compared to different versions of the java plugin sidecar.

Using an init container adds an extra step to the workspace bootstrap. 
Using an init container makes the meta.yaml harder to write and read.
These are costs that we can accept to pay if it has value for the user. But if 99% of the time users will just use the default target JDK is it really worth it? Shouldn't we just document how to use init containers to inject user-provided target JDKs and use sidecars for the default ones?

In your example you are defining the init Container in the devfile rather than [in the meta.yaml](https://github.com/eclipse/che-plugin-registry#plugin-meta-yaml-structure). Is that just for PoC purposes? Because if it's not in a `meta.yaml` it won't be reusable in other devfiles.

**Comment by tsmaeder** [comment]
I don't think the argument that the complication of writing a devfile is really a problem: we have to have the init container, whether we have it in the devfile or in a meta.yml. If _we_ write the devfile, complication is not a problem: if the users bring their own jdk, writing a devfile with an init container is no more nor less complicated than writing a meta.yml with an init container. (slightly less even, since they see them all the time).
What we'll end up with will be 1 Java plugin and n devfiles where n is the number of JDK's we want to support. That is important in order to reduce unnecessary maintenance and testing overhead.
Could you describe the alternative solution you have in mind?

**Comment by tsmaeder** [comment]
Also, for any m+n solution (as oppposed to m*n), we'll need an additional container in the workspace, whether it be a plugin per JDK version or an init container referenced in devfile. Copying the JDK is cheap: cf. the numbers above.

**Comment by l0rd** [comment]
> What we'll end up with will be 1 Java plugin and n devfiles where n is the number of JDK's we want to support.

User that need to create their custom java stack BUT are ok to use a default JDKs won't have an easy way to do that: they still will need to use the initcontainer mechanism as if they were including a "user-provided" JDK. And imo this is what 99% of users will want to do.

> Could you describe the alternative solution you have in mind? How would you reuse the init container?

If you include the initContainers mechanism in a meta.yaml you will be able to publish it in the plugin registry and user will be able to include the plugin in their devfiles.

**Comment by tsmaeder** [comment]
> Users that need to create their custom java stack BUT are ok to use a default JDKs

Why would they want a custom stack if they use the default JDK's? And they would just copy one of the provided devfiles. Simples!

**Comment by l0rd** [comment]
Even if they use the same jdk as in our getting started they may want to add some tools, change the projects, change commands...

**Comment by scela** [comment]
@ericwill wdyt ^?

**Comment by ericwill** [comment]
IMO we should steer users towards writing devfiles -- of varying complexities. This means thoroughly documenting as many workflows as possible along with their configurations. Of course, adding more devfile samples will also help here -- if we can cover 90% or more of all use cases with good documentation/samples, then we are doing a good job.

I think `initContainers` in the devfile is the way forward, for a couple of reasons:
* It opens up the ability for advanced users (corporate use cases, etc.) to define some nice flows to satisfy specific needs
* Other language stacks may benefit from this as well -- Java isn't the only language where the development environment is limited to the tooling environment
* Devfiles are like recipes for a reproducible Che environment: it's one of the main value-adds for using Che and we should encourage users to write them. If it's a question of difficulty we can definitely mitigate it on our end by improving docs, more samples, etc.
* Plugin maintainability is easier
* Performance doesn't seem to be adversely affected

Having users write `meta.yaml` files for plugins is something we should avoid -- once we have proper plugin automation in place, user involvement with writing `meta.yaml` files should be a thing of the past.


> Even if they use the same jdk as in our getting started they may want to add some tools, change the projects, change commands...

Maybe I am misunderstanding, but shouldn't this happen in the devfile regardless? From my POV:
1. The plugin registry is a static list of available tools (plugins)
2. A devfile defines a way to create a workspace using those pre-defined tools, with configuration options for said tools being defineable/configurable in the devfile

We should focus on making `2.` easy, while still allowing it to be powerful for advanced users.

**Comment by l0rd** [comment]
@ericwill `initContainers` are the way forward when users want to bring their own JDK. 

But why do you want users to specify an `initContainer`

```yaml
# using init container
---
components:
  - id: redhat/java11/latest
    type: chePlugin
  - referenceContent: |
      apiVersion: v1
      kind: Pod
      metadata:
        name: ws
      spec:
        initContainers:
          - name: init-myservice
            image: adoptopenjdk/openjdk11-openj9
            command: ['sh', '-c', "echo starting && cp -R /opt/java/openjdk/ /tmp/jdk && echo done && sleep 20"]
            volumeMounts: 
              - name:  remote-endpoint
                mountPath: /tmp/jdk
    type: kubernetes
```

when we can package that for them:

```yaml
# including jdk8 in the sidecar
---
components:
  - id: redhat/java8/latest
    type: chePlugin
```

And when I say "we can package" I mean that using the `vscode-extensions.json` format we could reference multiple Dockerfiles for the same vscode-extension:

```json
{
   "repository": "https://github.com/redhat-developer/vscode-java",
   "checkout": "v0.57.0",
   "dockerfiles": [
       "./dockerfiles/redhat-developer.vscode-java/Dockerfile.java8",
       "./dockerfiles/redhat-developer.vscode-java/Dockerfile.java11"
    ]
}
```

**Comment by l0rd** [comment]
@ericwill beyond the 14 extra lines needed in EVERY java devfile you also need to consider that the startup of the workspace will be slowed down by 
- the pull of the `initContainer` image 
- the copy of the JDK. 
Two operations that are NOT needed in the sidecar way.

And all of that for what? What is the benefit? I cannot see one honestly.

**Comment by ericwill** [comment]
Well there is the maintainability overhead. Using `initContainer` we would have to maintain only one Java plugin with one sidecar, and for other versions users bring their own JDK. We can facilitate this of course by providing pre-made devfile samples for some of the popular JDK's (11, 14, etc. come to mind) all while just maintaining the one plugin and one sidecar.

Using the `meta.yaml` proposal, we'd have to maintain _n_ number of sidecar images, as well as keep some number of Java plugins up-to-date and tested on each of those sidecars. Even with the automation it becomes a headache if we support more than two JDK versions.

> @ericwill beyond the 14 extra lines needed in EVERY java devfile you also need to consider that the startup of the workspace will be slowed down by
> 
>     * the pull of the `initContainer` image
> 
>     * the copy of the JDK.
>       Two operations that are NOT needed in the sidecar way.

The pulling I can see as a performance hit, can we optimize it somehow? For our pre-made devfile samples, can't we make use of the kubernetes image puller to cache the JDK images?

As for the copying of the JDK, from my understanding it's pretty cheap (@tsmaeder could you confirm)?

**Comment by l0rd** [comment]
Ok as I see it either we stop supporting jdt.ls with JDKs other than v11 so that we have less things to maintain or we continue supporting jdt.ls with popular JDKs so that we make users life easier.

But providing some sample devfiles with popular JDKs and NOT providing pre-packaged plugins we will have worst of both worlds! We will have to test jdt.ls against all JDKs AND users will need to write 14 extra lines of yaml to use those JDKs in their devfiles.

**Comment by tsmaeder** [comment]
@ericwill @l0rd  copy of the JDK is around 1.5 seconds on openshift.io, as mentioned further up in this issue. Also, if we provide a different JDK as a container, we will _always_ have to pull that container at some time. It doesn't matter if that container is referenced in a meta.yml or a devfile. So the net impact of the approach laid out above is ~1.5 seconds per workspace start. 

@l0rd if putting init containers into a workspace is so verbose, why not simplify it: we control the devfile syntax and can add any syntactic sugar we want. The 14 lines are there because I have to use the "kubernetes" type component and use the raw kube syntax for an init container. We could simplify that to something as simple as:

```
- initContainer: adoptopenjdk/openjdk11-openj9
  command: ['cp -R /opt/java/openjdk/ /tmp/jdk]
``` 

Also, we won't have to test the Java plugin against any JDK's. We'll always run jdt.ls on the same JDK. What we _will_ have to test is that the examples in the devfiles work as intended, but that is a much smaller set of things that could go wrong.

Also, we would not have to maintain a set of images for the various JDK's. We would just use the stock JDK images. 

I guess where we really differ is whether it is desirable to have users bring their own JDK to compile and debug against. I am convinced that people want exactly that. After all, one of the selling points of both containers in general and the Che 7 workspace model is that the environment in development is the same as the runtime model.

> Ok as I see it either we stop supporting jdt.ls with JDKs other than v11 so that we have less things to maintain or we continue supporting jdt.ls with popular JDKs so that we make users life easier.

No, that is not the alternative: we just support other _target_ JDK's via devfiles instead of making a custom plugin for each JDK version we support.

**Comment by l0rd** [comment]
> @ericwill @l0rd copy of the JDK is around 1.5 seconds on openshift.io, as mentioned further up in this issue. Also, if we provide a different JDK as a container, we will _always_ have to pull that container at some time. It doesn't matter if that container is referenced in a meta.yml or a devfile. So the net impact of the approach laid out above is ~1.5 seconds per workspace start.

We do not have to pull an init-container if the target JDK is embedded in jdt.ls sidecar. 

And we need a really good reason to accept to add 1.5s to the workspace startup process! We are trying hard to reduce seconds, adding even a fraction of second for no good reason is NOT acceptable.

> @l0rd if putting init containers into a workspace is so verbose, why not simplify it: we control the devfile syntax and can add any syntactic sugar we want.

That's a good point. I agree, we should do that for user provided JDKs. There are still 2 extra lines though.

> Also, we would not have to maintain a set of images for the various JDK's. We would just use the stock JDK images.

But that's what we are already doing. We are already leveraging 3rd party images rather than creating ours. The problem is that even if we use 3rd party images those get updated and we need to verify that the updates don't break our getting started. We have hundreds of 3rd party artifacts to maintain and we won't get rid of that with your proposal.  

**To summarize it is not clear what's the maintaince burden that you are experiencing today and how you are going to get rid of it. Bucause at the end that's the only good reason that would justify the extra costs (startup time and a more yaml lines in devfiles).**

**Comment by tsmaeder** [comment]
So in your world, how would we support users that want to use OpenJ9? The only way I see with your proposal is to add plugins for Java, Quarkus etc. for all versions of OpenJ9, as well. That is a pretty big maintenance burden, in my book.
Also, even the fact that we have to maintain plugins for the select few JDK versions we have now is something that we should get away from: it was a workaround for when jdt.ls could not separate the target JDK from the runtime. Apart from the maintenance, it makes no sense from the user's point of view, or have you ever wondered if you want to download Eclipse for Java 11 or Eclipse for Java 8?

**Comment by l0rd** [comment]
> So in your world, how would we support users that want to use OpenJ9? The only way I see with your proposal is to add plugins for Java, Quarkus etc. for all versions of OpenJ9, as well. That is a pretty big maintenance burden, in my book.

This is the user provided JDK use case right? And as I think I have mentioned in all the comments above I am +1 to use your proposal for those use cases. That makes perfect sense.

With regards to vsx dependencies: let's consider that we have [this pending issue ](https://github.com/eclipse/che/issues/15966) that would allow to merge all vsx in one container.

**Comment by ericwill** [comment]
> **To summarize it is not clear what's the maintaince burden that you are experiencing today and how you are going to get rid of it. Because at the end that's the only good reason that would justify the extra costs (startup time and a more yaml lines in devfiles).**

The cost we are trying to avoid is maintaining multiple plugins and their respective sidecars. Right now we need to maintain two plugins and two sidecars, just for the Java plugin. When we want to start supporting newer JDKs this cost will increase.

By the way I am not talking about other tools/containers like the jdk-maven images, those are separate and nothing changes there -- I am specifically talking about vscode-java/jdt.ls and its sidecar. We will still have to validate different Java versions and samples in the devfiles we provide -- this proposal isn't trying to change that.

Consider the following: let's say today we wanted to add Java 14 support. With the current mechanism we would need to:
* Add a new plugin to the registry
* Add a new sidecar to the che-sidecar-java repo
* Maintain both for _x_ number of sprints
* Add a new Java 14 devfile with code samples, Java 14 maven image, and other tools

With the `initContainer` approach the same scenario would look like this:
* Add a new Java 14 devfile with code samples, Java 14 maven image and other tools, include an `initContainer` for the Java 14 JDK and point jdt.ls to it

In the second scenario the only maintenance associated with the Java plugin is maintaining one plugin in the registry, and one sidecar. The Java version there is whatever jdt.ls' preferred runtime is -- Java 11 in this case as its the minimum. Adding support for new Java versions then only becomes a question of adding a new devfile with samples, and testing it (which is already required today, as you pointed out), and adding an `initContainer` and pointing jdt.ls to whatever the target JDK is.

Also +1 to making the devfile syntax easier -- I think 2 lines is much more manageable than 14.

**Comment by benoitf** [comment]
Hi, IMHO the problem with saying "In the second scenario the only maintenance associated with the Java plugin is maintaining one plugin in the registry, and one sidecar",

is that you push maintenance of JDK as well to users.

They don't need to only select "plugin Java11, Java12, Java13, Java14", they do need to specify more stuff in each of their devfile

**Comment by ericwill** [comment]
> Hi, IMHO the problem with saying "In the second scenario the only maintenance associated with the Java plugin is maintaining one plugin in the registry, and one sidecar",
> 
> is that you push maintenance of JDK as well to users.
> 
> They don't need to only select "plugin Java11, Java12, Java13, Java14", they do need to specify more stuff in each of their devfile

True, maybe we can improve it? Contribute a plugin that allows for easy switching of JDK target environments through the UI:
* Select from a list of existing JDK images (8, 11, 14, etc.)
* Provide a link to a JDK container

Once a selection is made we could generate the appropriate lines and add them to the devfile. Or maybe that's not possible (it's just a thought)?

**Comment by l0rd** [comment]
> * Add a new plugin to the registry 

That's not needed with the new vscode-extensions.json as described [here](https://github.com/eclipse/che/issues/16851#issuecomment-640773205)
```json
{
   "repository": "https://github.com/redhat-developer/vscode-java",
   "checkout": "v0.57.0",
   "dockerfiles": [
       "./dockerfiles/redhat-developer.vscode-java/Dockerfile.java11",
       "./dockerfiles/redhat-developer.vscode-java/Dockerfile.java14"
    ]
}
```

> * Add a new sidecar to the che-sidecar-java repo

That's equivalent to "add a new initContainer" in the new approach

> * Maintain both for _x_ number of sprints

You need that in the new approach as well: maintain initcontainer + plugin

> * Add a new Java 14 devfile with code samples, Java 14 maven image, and other tools

You need that in the new approach as well

**Comment by ericwill** [comment]
> > * Maintain both for _x_ number of sprints
> 
> You need that in the new approach as well: maintain initcontainer + plugin

AIUI the `initContainer` does not require active maintenance, is it not just an upstream JDK image? Like [openjdk](https://hub.docker.com/_/openjdk), for example.

**Comment by tsmaeder** [comment]
@l0rd  could you describe where that vscode-extensions.json and how the "dockerfiles" mechanism would work? It's the first time I hear of it.

**Comment by l0rd** [comment]
> AIUI the `initContainer` does not require active maintenance, is it not just an upstream JDK image? Like [openjdk](https://hub.docker.com/_/openjdk), for example.

@ericwill In both cases, in an `initContainer` and in a `Dockerfile`, you are copying a JDK into an image. The only difference is when you do that (at runtime or at build time). Hence you are only moving the maitainance problem somewhere else. To illustrate my reasoning here is a Dockerfile that does the same copy as the initcontainer + volume mechanism above:

```Dockerfile
FROM adoptopenjdk/openjdk11-openj9 as jdk

FROM quay.io/eclipse/che-sidecar-java:11-f76ca45
COPY --from=jdk /opt/java/openjdk/ /remote-endpoint/openjdk
```

Let me be clear: pre-built sidecars are better for the 2-3 JDKs that we support BUT the initcontainer approach is much better for users that want to user their own JDK because they only need to add a few lines to their `devfile` instead of 1) do a build of a sidecar image 2) push it to an OCI registry 3) build a new plugin registry image 4) rollout the new plugin registry. In that case the extra costs of the `initContainer` approach are justified.

> @l0rd could you describe where that vscode-extensions.json and how the "dockerfiles" mechanism would work? It's the first time I hear of it.

@tsmaeder this is described in #17029 that is part of this epic #15819.

**Comment by nickboldt** [comment]
Hey, there, long time listener, first time caller. 

I have a couple questions:

a) is it an existing customer need that we have the ability to easily (ie., UI toggle) between installed/available JDKs? Since I almost never write java anymore, and since old JDKs are just ancient history, I don't know if a Che or CRW customer would legit care about having more than 1 JDK available in the IDE, like one can with desktop Eclipse.

b) the ability to disconnect the JDK from the plugin sidecar would allow us to WAY MORE EASILY support new arches like s390x (Z for Linux) and ppc64le (IBM Power), as we would NOT have to have two essentially-identical images:

* openjdk8/maven, python3/pip, node/npm
* openjdk11/maven, quarkus, gradle

and

* openj9-8/maven, python3/pip, node/npm
* openj9-11/maven, quarkus, gradle

Instead we would be able to reuse existing Red hat Container Catalogue openJDK & openJ9 images as is, and would also therefore benefit from faster CVE fixes from that team independent of the Che/CRW development/productization efforts.

(Turns out that second one wasn't a question, but rather an endorsement of the idea of JDK decoupling from the sidecar.)

**Comment by nickboldt** [comment]
I've heard this won't get into 7.21 / 7.20.2 so I'll set a 7.22 milestone for now just so it's got a target to work toward

**Comment by ericwill** [comment]
> I've heard this won't get into 7.21 / 7.20.2 so I'll set a 7.22 milestone for now just so it's got a target to work toward

Thanks Nick. I've added the sub-issues in the OP of this thread.

**Comment by nickboldt** [comment]
Would using https://sdkman.io/ get us closer to having multiple JDK support inside Che/CRW in the same container (including airgapped support) ?

**Comment by tsmaeder** [comment]
> if an airgapped customer wants something else like 11 or 15, they'd have to figure out how to break thru the airgap to download the other SDK rpms into their environment.

That's exactly the behaviour that you want to prevent with an airgap. So no, I don't think that is going to fly. The solution presented above neatly takes care of the problems, we just need to go ahead and implement it.
